graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DataFetcherInvoker.kt

```diff
@@ -113,19 +123,40 @@ class DataFetcherInvoker internal constructor(
                 kFunc.callSuspendBy(argsByName)
             }.onErrorMap(InvocationTargetException::class.java) { it.targetException }
         }
-        return try {
-            if (completableFutureWrapper.shouldWrapInCompletableFuture(kFunc)) {
-                completableFutureWrapper.wrapInCompletableFuture { kFunc.callBy(argsByName) }
-            } else {
+        return if (completableFutureWrapper.shouldWrapInCompletableFuture(kFunc)) {
+            completableFutureWrapper.wrapInCompletableFuture { kFunc.callBy(argsByName) }
+        } else {
+            try {
                 kFunc.callBy(argsByName)
+            } catch (exc: Exception) {
+                handleReflectionException(exc)
             }
-        } catch (ex: Exception) {
-            ReflectionUtils.handleReflectionException(ex)
         }
     }
 
     private fun formatArgumentError(param: MethodParameter, message: String): String {
         return "Could not resolve parameter [${param.parameterIndex}] in " +
             param.executable.toGenericString() + if (message.isNotEmpty()) ": $message" else ""
     }
+
+    /**
+     * Handle the given reflection exception.
+     *
+     * Variant of [ReflectionUtils.handleReflectionException] that allows checked exceptions
+     * to propagate, but handles [NoSuchMethodException], [IllegalAccessException], and [InvocationTargetException]
+     * the same way as that helper does; the main difference is that this method that this method will never throw
+     * [UndeclaredThrowableException].
+     */
+    private fun handleReflectionException(exc: Exception): Nothing {
+        if (exc is NoSuchMethodException) {
+            throw IllegalStateException("Method not found: ${exc.message}")
+        }
+        if (exc is IllegalAccessException) {
+            throw IllegalStateException("Could not access method or field: ${exc.message}")
+        }
+        if (exc is InvocationTargetException) {
+            throw exc.targetException
+        }
```

Comment:
Don't we need to handle RunTimeException case as they do in ReflectionUtils?

Reply:
Runtime exceptions will just get rethrown, which is the same behavior as ReflectionUtils; the reason they have special handling of RuntimeException is that they need to filter out RuntimeException so they can wrap checked Exceptions in UndeclaredThrowableException. Since we don't want that behavior at all we don't need the extra logic.